src: add support for nogc types via `NogcEnv` (#1514)
nodejs/node-addon-api: doc/finalization.md

# Finalization

Various node-addon-api methods accept a templated `Finalizer finalizeCallback`
parameter. This parameter represents a native callback function that runs in
response to a garbage collection event. A finalizer is considered a _basic_
finalizer if the callback only utilizes a certain subset of APIs, which may

> provide optimizations, improved execution, or other benefits.

Reviewer: ```suggestion
provide more efficient memory management, optimizations, improved execution, or other benefits.
```

I think we should explicitly point out memory management because that's what caused us to go down this path in the first place.

Author: Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

In general, it is best to use basic finalizers whenever possible (eg. when
access to JavaScript is _not_ needed).

## Finalizers

The callback takes `Napi::Env` as its first argument:

### Example

```cpp
Napi::External<int>::New(Env(), new int(1), [](Napi::Env env, int* data) {
  env.RunScript("console.log('Finalizer called')");
  delete data;
});
```

## Basic Finalizers

Use of basic finalizers may allow the engine to perform optimizations when
scheduling or executing the callback. For example, V8 does not allow access to
the engine heap during garbage collection. Restricting finalizers from accessing
the engine heap allows the callback to execute during garbage collection,

> providing a chance to free native memory on the current tick.

Reviewer 2: I think it would be good to just say "eagerly" being GC-ed.

```suggestion
providing a chance to free native memory eagerly.
```

Author: Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).

In general, APIs that access engine heap are not allowed in basic finalizers.

The callback takes `Napi::BasicEnv` as its first argument:

### Example

```cpp
Napi::ArrayBuffer::New(
    Env(), data, length, [](Napi::BasicEnv /*env*/, void* finalizeData) {
      delete[] static_cast<uint8_t*>(finalizeData);
    });
```

## Scheduling Finalizers

In addition to passing finalizers to `Napi::External`s and other Node-API

> constructs, use `Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer)` to

Reviewer: ```suggestion
constructs, `Napi::BasicEnv::PostFinalize(Napi::BasicEnv, Finalizer)` can be used to
```

We should use passive voice.

Author: Addressed in [195ec28](https://github.com/nodejs/node-addon-api/pull/1514/commits/195ec28a7b1dcfe571c6df5f8a2e1bb991126ff1).
